### PR TITLE
refactor ndarray function parsing & add feed_dict to bind for autograd

### PR DESCRIPTION
**Pull request by piiswrong**

### CODE REVIEW

File: src/operator/tensor/sample_op.cu [hunk 1]
```diff
@@ -67,10 +67,10 @@ void SampleNormal_<gpu>(const nnvm::NodeAttrs& attrs,
   }
 }
 
-NNVM_REGISTER_OP(uniform)
+NNVM_REGISTER_OP(random_uniform)
```
Comment (piiswrong) [inline src/operator/tensor/sample_op.cu]: this was failing on my machine but seems was fine on test server. I fixed it to be safe.

File: src/operator/tensor/sample_op.h [hunk 2]
```diff
@@ -105,12 +105,12 @@ struct SampleGammaParam : public dmlc::Parameter<SampleGammaParam> {
 };
 
 struct SampleExponentialParam : public dmlc::Parameter<SampleExponentialParam> {
-  float lambda;
+  float lam;
```
Comment (piiswrong) [inline src/operator/tensor/sample_op.h]: @asmushetzel lambda is a python keyword and cannot be used in frontend.

File: src/operator/tensor/sample_op.cu (hunk 1 above)
Comment (asmushetzel) [inline src/operator/tensor/sample_op.cu]: Thanks for catching.

File: python/mxnet/contrib/autograd.py [hunk 3]
```diff
@@ -10,15 +10,70 @@
...
+    return bool(prev.value)
+
+
+class TrainingStateScope(object):
+    """Scope for managing training state.
+
+    Example::
+        with TrainingStateScope(True):
+            y = model(x)
+            compute_gradient([y])
```
Comment (ZihengJiang) [inline python/mxnet/contrib/autograd.py]: I think `compute_gradient` should not be put in the training scope? The example may need to be changed.

File: python/mxnet/contrib/autograd.py [hunk 4]
```diff
@@ -10,15 +10,70 @@
...
+
+    def __enter__(self):
+        self._prev = set_is_training(self._enter_state)
+
+    def __exit__(self, ptype, value, trace):
+        if self._prev != self._enter_state:
+            set_is_training(self._prev)
+
+
+def train():
```
Comment (ZihengJiang) [inline python/mxnet/contrib/autograd.py]: `train` and `test` sound too generalized, maybe consider `train_scope` and `test_scope`

**Pull request merged**